Disable storing of application data in case of WANT_WRITE error

steweg opened the pull request:
This patch disables storing of application data in case of WANT_WRITE error, in order to allow applications using wolfSSL to use the same logic, when compered to EAGAIN/EWOULDBLOCK from kernel sockets. Currently the difference is that wolfSSL stores the packet within its internal buffer and expects that application will resend exactly the same data during the next call. If application doesn't do that and by accident the send application data are having the same length as stored packet, the stored packet is sent instead and the application data are not send at all, while the result of sending operation is success, therefore the application is not aware and the TLS connection might be broken for good.

This patch in case of WANT_WRITE error simulates correct sending from internal buffer perspective and as such the application can correctly detect what was or wasn't sent and behave accordingly

As I don't known what was the reason for such design/behavior in the past, please review from your side, if it would not have some additional side-effects that I might have missed.

ejohnstown commented:
It might not be clear from the manual, but if a wolfSSL_write() returns the error WANT_WRITE, you must call wolfSSL_write() again with the same data. It uses that to tell how much it sent.

Thank you for the contribution, but we cannot accept it.

steweg commented:
Ok, understood. Can I at least ask for explanation to be added into documentation, why the application must send the same data?

The main reason why I introduced this patch was that I was working on application, which used it's own application protocol that didn't require packet order or guarantee of delivery, while one of possible underlying transports was TLS tunnels based on wolfSSL. We also added our own outgoing message prioritization. But this is simply not possible to be achieved thanks to requirement to resend the same data first. Consider the scenario when while socket is busy, application simply gets request to send more urgent data out, so immediately when socket becomes available, it will try to send more urgent data first, and then if needed re-transmit the failed data. This is easily possible using basic TCP sockets, but not over TLS ones. And what is even worse, there is no info, why this is a must. 

Thanks for understading

ejohnstown commented:
wolfSSL_write() uses the length of the data passed into it and the amount of data encoded and already sent to tell that it finished sending the message. I always looked at wolfSSL_write() as the border for the application data. Once you give it to wolfSSL_write(), it is decorating it before giving it to the socket. So in your case, the prioritization comes before the socket send() in the non-encrypted version, and should happen before the wolfSSL_write() in the encrypted version.

ejohnstown commented:
OK, I understand what you are trying to do now.

If the write blocks and you drop the pending send data, the TLS session will end up in an inconsistent state. Any future packets you send (except for maybe AEAD packets) will fail the MAC check. TLS doesn't send the sequence number in the packet, each endpoint increments the sequence number when sending or receiving a record. The sequence number is part of the MAC.

When you call wolfSSL_write(), first space is allocated in the output buffer based on the size of the data you want to send and the overhead for the TLS record. The TLS record header is inserted into the output buffer. Your data to send is encrypted from your buffer into the output buffer. The MAC code is added as a part of this process. During the MAC or AEAD encrypt process, the TLS implicit sequence number is incremented. Then the send is attempted.

steweg commented:
Well not really, the tcp stream will be consistent as if you get EAGAIN error, nothing was sent to the remote side, so it doesn’t increment counter on receiver side, nor sender side cannot increment anything as it was not send at all and must return the state of TLS session as it was before otherwise the counters will be incremented one more time during next send attempt. If counters are not send to receiver, the receiver will not be able to decrypt or verify MAC properly. So either the TLS must send the counter or TLS must return counter back to previous state if sending fails. Anyway there is still no need to send exactly the same data as before.

I have tried to use this patch code for some time and the TLS session is always able to survive this state without any issues

ejohnstown commented:
I agree the TCP stream will be consistent if you get the EAGAIN error. The TLS stream will not be consistent.

wolfSSL_write() will increment the TLS sequence number when it calculates the HMAC on the record. Then it encrypts the record. Then it tries to send record. If the record is dropped on EAGAIN, the peer will not get that record and will not increment its accounting of this sequence number. The next message will have a new sequence number, the peer will not verify the MAC correctly.

"So either the TLS must send the counter" is outside the scope of TLS.  "[Or] TLS must return counter back to previous state if sending fails," your patch is incomplete.

This sounds like a useful option. Unfortunately, we cannot accept this PR at this time.
